src/protogen/GrpcTools.cs

```diff
@@ -0,0 +1,146 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace ProtoBuf
+{
+    internal static class GrpcTools
+    {
+        enum GrpcMode
+        {
+            List,
+            Get,
+        }
+
+        internal static async Task<int> ExecuteAsync(string modeString, string uri, string serviceName,
+            CodeGenerator codegen, string outPath, Dictionary<string,string> options)
+        {
+            
+            if (!Enum.TryParse<GrpcMode>(modeString, true, out var mode))
+            {
+                Console.Error.WriteLine($"Unknown gRPC mode: '{modeString}'");
+                return 1;
+            }
+            if (string.IsNullOrWhiteSpace(outPath))
+            {
+                Console.Error.WriteLine($"Missing output directive; please specify --csharp_out etc");
+                return 1;
+            }
+
+            switch (mode)
+            {
+                case GrpcMode.List:
+                    Console.WriteLine($"Requesting gRPC service directory from '{uri}'...");
+                    break;
+                case GrpcMode.Get:
+                    Console.WriteLine($"Requesting gRPC '{serviceName}' service from '{uri}'...");
+                    break;
+                default:
+                    Console.Error.WriteLine($"Unexpected mode: {mode}");
+                    return 1;
+            }
+            int errorCount = 0;
+            GrpcClientFactory.AllowUnencryptedHttp2 = true;
+            using var channel = GrpcChannel.ForAddress(uri);
+            var service = channel.CreateGrpcService<IServerReflection>();
+
+            FileDescriptorSet set = null;
+            int services = 0;
+            try
+            {
+                await foreach (var reply in service.ServerReflectionInfoAsync(GetRequest()))
+                {
+                    switch (reply.MessageResponseCase)
+                    {
+                        case ServerReflectionResponse.MessageResponseOneofCase.ListServicesResponse:
+                            foreach (var availableService in reply.ListServicesResponse.Services)
+                            {
+                                services++;
+                                Console.WriteLine($"- {availableService.Name}");
+                            }
+                            break;
+                        case ServerReflectionResponse.MessageResponseOneofCase.FileDescriptorResponse:
+                            {
+                                var file = reply.FileDescriptorResponse;
+                                if (file is null) continue;
+                                foreach (byte[] payload in file.FileDescriptorProtoes)
+                                {
+                                    var proto = Serializer.Deserialize<FileDescriptorProto>(new Span<byte>(payload));
+                                    proto.IncludeInOutput = true; // have to assume all
```

Comment:
You probably need to set import (`foreach (var dependency in proto.Dependencies) { proto.AddImport(dependency, true, default); }`) as well. Additionally, should "well known types" be excluded from output?